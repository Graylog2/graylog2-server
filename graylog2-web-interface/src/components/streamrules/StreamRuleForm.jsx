import PropTypes from 'prop-types';
import React from 'react';
import { Col } from 'react-bootstrap';

import { Input } from 'components/bootstrap';
import BootstrapModalForm from 'components/bootstrap/BootstrapModalForm';
import { TypeAheadFieldInput } from 'components/common';
import { DocumentationLink } from 'components/support';
import DocsHelper from 'util/DocsHelper';
import Version from 'util/Version';
import FormsUtils from 'util/FormsUtils';

import HumanReadableStreamRule from 'components/streamrules//HumanReadableStreamRule';

class StreamRuleForm extends React.Component {
  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    streamRule: PropTypes.object,
    streamRuleTypes: PropTypes.array.isRequired,
    title: PropTypes.string.isRequired,
  };

  static defaultProps = {
    streamRule: { field: '', type: 1, value: '', inverted: false, description: '' },
  };

  state = this.props.streamRule;
  FIELD_PRESENCE_RULE_TYPE = 5;
  ALWAYS_MATCH_RULE_TYPE = 7;
  modal = undefined;

  _resetValues = () => {
    this.setState(this.props.streamRule);
  };

  _onSubmit = () => {
    if (this.state.type === this.ALWAYS_MATCH_RULE_TYPE) {
      this.state.field = '';
    }
    if (this.state.type === this.FIELD_PRESENCE_RULE_TYPE || this.state.type === this.ALWAYS_MATCH_RULE_TYPE) {
      this.state.value = '';
    }
    this.props.onSubmit(this.props.streamRule.id, this.state);
    this.modal.close();
  };

  _formatStreamRuleType = (streamRuleType) => {
    return (
      <option key={`streamRuleType${streamRuleType.id}`}
              value={streamRuleType.id}>{streamRuleType.short_desc}</option>
    );
  };

  open = () => {
    this._resetValues();
    this.modal.open();
  };

  close = () => {
    this.modal.close();
  };

  handleChange = (event) => {
    const change = {};
    change[event.target.name] = FormsUtils.getValueFromInput(event.target);
    this.setState(change);
  };

  render() {
    const { field, type, value, inverted, description } = this.state;

    const streamRuleTypes = this.props.streamRuleTypes.map(this._formatStreamRuleType);
    const fieldBox = (String(type) !== String(this.ALWAYS_MATCH_RULE_TYPE) ?
      <TypeAheadFieldInput id="field-input" type="text" required label="Field" name="field" defaultValue={field} onChange={this.handleChange} autoFocus /> : '');
    const valueBox = (String(type) !== String(this.FIELD_PRESENCE_RULE_TYPE) && String(type) !== String(this.ALWAYS_MATCH_RULE_TYPE) ?
      <Input id="Value" type="text" required label="Value" name="value" value={value} onChange={this.handleChange} /> : '');
    return (
      <BootstrapModalForm ref={(c) => { this.modal = c; }}
                          title={this.props.title}
                          onSubmitForm={this._onSubmit}
                          submitButtonText="Save"
                          formProps={{ id: 'StreamRuleForm' }}>
        <div>
          <Col md={8}>
            {fieldBox}
            <Input id="Type" type="select" required label="Type" name="type" value={type} onChange={this.handleChange}>
              {streamRuleTypes}
            </Input>
            {valueBox}
            <Input id="Inverted" type="checkbox" label="Inverted" name="inverted" checked={inverted} onChange={this.handleChange} />

            <Input id="Description" type="textarea" label="Description (optional)" name="description" value={description} onChange={this.handleChange} />

            <p>
              <strong>Result:</strong>
              {' '}
              <HumanReadableStreamRule streamRule={this.state} streamRuleTypes={this.props.streamRuleTypes} />
            </p>
          </Col>
          <Col md={4}>
            <div className="well well-sm matcher-github">
              The server will try to convert to strings or numbers based on the matcher type as good as it
              can.

              <br /><br />
              <i className="fa fa-github" />
              <a href={`https://github.com/Graylog2/graylog2-server/tree/${Version.getMajorAndMinorVersion()}/graylog2-server/src/main/java/org/graylog2/streams/matchers`}
                 target="_blank"> Take a look at the matcher code on GitHub
              </a>
              <br /><br />
              Regular expressions use Java syntax. <DocumentationLink page={DocsHelper.PAGES.STREAMS}
                                                                      title="More information"
                                                                      text={<i className="fa fa-lightbulb-o" />} />
            </div>
          </Col>
        </div>
      </BootstrapModalForm>
    );
  }
}

export default StreamRuleForm;
