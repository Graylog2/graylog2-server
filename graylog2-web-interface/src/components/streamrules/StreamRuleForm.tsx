/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import PropTypes from 'prop-types';
import React from 'react';

import { Col, Well } from 'components/graylog';
import { BootstrapModalForm, Input } from 'components/bootstrap';
import { Icon, TypeAheadFieldInput } from 'components/common';
import { DocumentationLink } from 'components/support';
import HumanReadableStreamRule from 'components/streamrules//HumanReadableStreamRule';
import CombinedProvider from 'injection/CombinedProvider';
import connect from 'stores/connect';
import DocsHelper from 'util/DocsHelper';
import Version from 'util/Version';
import * as FormsUtils from 'util/FormsUtils';
import { Store } from 'stores/StoreTypes';

const { InputsStore, InputsActions } = CombinedProvider.get('Inputs');

type StreamRule = {
  type: number,
  field: string,
  value: string,
  id?: string,
  inverted: boolean,
  description: string,
};

type StreamRuleType = {
  id: number,
  short_desc: string,
  long_desc: string,
  name: string,
};

type Props = {
  onSubmit: (streamRuleId: string | undefined | null, currentStreamRule: StreamRule) => void,
  streamRule: StreamRule,
  streamRuleTypes: Array<StreamRuleType>,
  title: string,
  inputs: Array<unknown>,
  onClose: () => void,
};

type State = {
  streamRule: StreamRule,
  error: string,
};

class StreamRuleForm extends React.Component<Props, State> {
  static defaultProps = {
    // eslint-disable-next-line react/default-props-match-prop-types
    streamRule: { field: '', type: 1, value: '', inverted: false, description: '' },
    // eslint-disable-next-line react/default-props-match-prop-types
    inputs: [],
    onClose: () => {},
  };

  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    streamRuleTypes: PropTypes.array.isRequired,
    title: PropTypes.string.isRequired,
    onClose: PropTypes.func,
  };

  FIELD_PRESENCE_RULE_TYPE = 5;

  ALWAYS_MATCH_RULE_TYPE = 7;

  MATCH_INPUT = 8;

  PLACEHOLDER_INPUT = 0;

  constructor(props) {
    super(props);

    this.state = {
      streamRule: props.streamRule,
      error: '',
    };
  }

  componentDidMount() {
    InputsActions.list();
  }

  _validateMatchInput = () => {
    const { streamRule: { value } } = this.state;

    if (String(value) === String(this.PLACEHOLDER_INPUT)) {
      this.setState({ error: 'Please choose an input' });

      return false;
    }

    this.setState({ error: '' });

    return true;
  };

  _onSubmit = () => {
    const { streamRule: currentStreamRule } = this.state;
    const { type } = currentStreamRule;
    const { streamRule, onSubmit } = this.props;
    const { onClose } = this.props;

    if (type === this.ALWAYS_MATCH_RULE_TYPE) {
      currentStreamRule.field = '';
      this.setState({ streamRule: currentStreamRule });
    }

    if (type === this.FIELD_PRESENCE_RULE_TYPE || type === this.ALWAYS_MATCH_RULE_TYPE) {
      currentStreamRule.value = '';
      this.setState({ streamRule: currentStreamRule });
    }

    if (String(type) === String(this.MATCH_INPUT)) {
      if (!this._validateMatchInput()) {
        return;
      }
    }

    onSubmit(streamRule.id, currentStreamRule);
    onClose();
  };

  _formatStreamRuleType = (streamRuleType) => (
    <option key={`streamRuleType${streamRuleType.id}`}
            value={streamRuleType.id}>
      {streamRuleType.short_desc}
    </option>
  );

  handleChange = (event) => {
    const { streamRule } = this.state;
    const updatedStreamRule = { ...streamRule };

    updatedStreamRule[event.target.name] = FormsUtils.getValueFromInput(event.target);

    if (event.target.name === 'type' && String(updatedStreamRule.type) === String(this.MATCH_INPUT)) {
      updatedStreamRule.value = String(this.PLACEHOLDER_INPUT);
    }

    this.setState({ streamRule: updatedStreamRule });
  };

  _formatInputOptions = (input) => (
    <option key={`input-${input.id}`} value={input.id}>
      {input.title} ({input.name})
    </option>
  );

  valueBox = () => {
    const { streamRule: { value, type }, error } = this.state;
    const { inputs } = this.props;

    switch (String(type)) {
      case String(this.FIELD_PRESENCE_RULE_TYPE):
      case String(this.ALWAYS_MATCH_RULE_TYPE):
        return '';

      case String(this.MATCH_INPUT): {
        const bsStyle = error && error.length > 0 ? 'error' : undefined;

        return (
          <Input bsStyle={bsStyle}
                 help={error}
                 id="Value"
                 type="select"
                 required
                 label="Value"
                 name="value"
                 value={value}
                 data-testid="input-selection"
                 onChange={this.handleChange}>
            <option value={this.PLACEHOLDER_INPUT}>Choose Input</option>
            {inputs.map(this._formatInputOptions)}
          </Input>
        );
      }

      default:
        return <Input id="Value" type="text" required label="Value" name="value" value={value} onChange={this.handleChange} />;
    }
  };

  fieldBox = () => {
    const { streamRule: { field, type } } = this.state;

    switch (String(type)) {
      case String(this.ALWAYS_MATCH_RULE_TYPE):
      case String(this.MATCH_INPUT):
        return '';
      default:
        return <TypeAheadFieldInput id="field-input" type="text" required label="Field" name="field" defaultValue={field} onChange={this.handleChange} autoFocus />;
    }
  };

  render() {
    const { streamRule } = this.state;
    const { type, inverted, description } = streamRule;
    const { streamRuleTypes: ruleTypes, title, onClose, inputs } = this.props;

    const streamRuleTypes = ruleTypes.map(this._formatStreamRuleType);
    const fieldBox = this.fieldBox();
    const valueBox = this.valueBox();

    return (
      <BootstrapModalForm title={title}
                          show
                          onCancel={onClose}
                          onModalClose={onClose}
                          onSubmitForm={this._onSubmit}
                          submitButtonText="Save"
                          formProps={{ id: 'StreamRuleForm' }}>
        <div>
          <Col md={8}>
            {fieldBox}
            <Input id="Type" data-testid="rule-type-selection" type="select" required label="Type" name="type" value={type} onChange={this.handleChange}>
              {streamRuleTypes}
            </Input>
            {valueBox}
            <Input id="Inverted" type="checkbox" label="Inverted" name="inverted" checked={inverted} onChange={this.handleChange} />

            <Input id="Description" type="textarea" label="Description (optional)" name="description" value={description} onChange={this.handleChange} />

            <p>
              <strong>Result:</strong>
              {' '}
              <HumanReadableStreamRule streamRule={streamRule} streamRuleTypes={ruleTypes} inputs={inputs} />
            </p>
          </Col>
          <Col md={4}>
            <Well bsSize="small" className="matcher-github">
              The server will try to convert to strings or numbers based on the matcher type as well as it can.

              <br /><br />
              <Icon name="github" type="brand" />&nbsp;
              <a href={`https://github.com/Graylog2/graylog2-server/tree/${Version.getMajorAndMinorVersion()}/graylog2-server/src/main/java/org/graylog2/streams/matchers`}
                 target="_blank"
                 rel="noopener noreferrer"> Take a look at the matcher code on GitHub
              </a>
              <br /><br />
              Regular expressions use Java syntax. <DocumentationLink page={DocsHelper.PAGES.STREAMS}
                                                                      title="More information"
                                                                      text={<Icon name="lightbulb" type="regular" />} />
            </Well>
          </Col>
        </div>
      </BootstrapModalForm>
    );
  }
}

type InputsStoreState = {
  inputs: Array<unknown>;
};

export default connect(StreamRuleForm,
  { inputs: InputsStore as Store<InputsStoreState> },
  ({ inputs }) => ({ inputs: inputs.inputs }));
