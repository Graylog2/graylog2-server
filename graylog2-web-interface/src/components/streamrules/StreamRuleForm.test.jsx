// @flow strict
import * as React from 'react';
import { cleanup, render } from 'wrappedTestingLibrary';
import StreamRuleForm from './StreamRuleForm';

jest.mock('components/common', () => ({
  // eslint-disable-next-line react/prop-types
  TypeAheadFieldInput: ({ defaultValue }) => (<div>{defaultValue}</div>),
  // eslint-disable-next-line react/prop-types
  Icon: ({ children }) => (<div>{children}</div>),
}));

describe('StreamRuleForm', () => {
  afterEach(() => {
    cleanup();
  });

  const streamRuleTypes = [
    { id: 1, short_desc: 'match exactly', long_desc: 'match exactly' },
    { id: 2, short_desc: 'match regular expression', long_desc: 'match regular expression' },
    { id: 3, short_desc: 'greater than', long_desc: 'greater than' },
    { id: 4, short_desc: 'smaller than', long_desc: 'smaller than' },
    { id: 5, short_desc: 'field presence', long_desc: 'field presence' },
    { id: 6, short_desc: 'contain', long_desc: 'contain' },
    { id: 7, short_desc: 'always match', long_desc: 'always match' },
    { id: 8, short_desc: 'match input', long_desc: 'match input' },
  ];

  const getStreamRule = (type = 1) => {
    return {
      id: 'dead-beef',
      type,
      field: 'field_1',
      value: 'value_1',
      inverted: false,
      description: 'description',
    };
  };

  it('should render an empty StreamRuleForm', () => {
    const container = render(
      <StreamRuleForm onSubmit={() => {}}
                      streamRuleTypes={streamRuleTypes}
                      title="Bach" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render an simple StreamRuleForm', () => {
    const container = render(
      <StreamRuleForm onSubmit={() => {}}
                      streamRule={getStreamRule()}
                      streamRuleTypes={streamRuleTypes}
                      title="Bach" />,
    );
    expect(container).toMatchSnapshot();
  });
});
