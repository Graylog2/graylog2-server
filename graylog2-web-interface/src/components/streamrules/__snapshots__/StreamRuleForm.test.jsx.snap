// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamRuleForm should render an empty StreamRuleForm 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    style="padding-right: 0px;"
  >
    <div
      aria-hidden="true"
    />
    <div
      aria-hidden="true"
      role="dialog"
    >
      <div
        class="fade modal-backdrop in"
      />
      <div
        class="fade Modal-nuree8-0 UjHmV in modal"
        role="dialog"
        style="display: block; padding-right: 0px;"
        tabindex="-1"
      >
        <div
          class="modal-dialog"
        >
          <div
            class="modal-content"
            role="document"
          >
            <div
              class="modal-header"
            >
              <button
                class="close"
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  Ã—
                </span>
                <span
                  class="sr-only"
                >
                  Close
                </span>
              </button>
              <h4
                class="modal-title"
              >
                Bach
              </h4>
            </div>
            <form
              data-testid="modal-form"
              id="StreamRuleForm"
            >
              <div
                class="modal-body"
              >
                <div
                  class="container-fluid"
                >
                  <div>
                    <div
                      class="col-md-8"
                    >
                      <div />
                      <div
                        class="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr form-group"
                      >
                        <label
                          class="ControlLabel-sc-1edmum5-0 hCskeX control-label"
                          for="Type"
                        >
                          Type
                        </label>
                        <span>
                          <select
                            class="FormControl-sc-1amoaox-0 fdBWEX form-control"
                            id="Type"
                            label="Type"
                            name="type"
                            placeholder=""
                            required=""
                            type="select"
                          >
                            <option
                              value="1"
                            >
                              match exactly
                            </option>
                            <option
                              value="2"
                            >
                              match regular expression
                            </option>
                            <option
                              value="3"
                            >
                              greater than
                            </option>
                            <option
                              value="4"
                            >
                              smaller than
                            </option>
                            <option
                              value="5"
                            >
                              field presence
                            </option>
                            <option
                              value="6"
                            >
                              contain
                            </option>
                            <option
                              value="7"
                            >
                              always match
                            </option>
                            <option
                              value="8"
                            >
                              match input
                            </option>
                          </select>
                          
                        </span>
                      </div>
                      <div
                        class="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr form-group"
                      >
                        <label
                          class="ControlLabel-sc-1edmum5-0 hCskeX control-label"
                          for="Value"
                        >
                          Value
                        </label>
                        <span>
                          <input
                            class="FormControl-sc-1amoaox-0 fdBWEX form-control"
                            id="Value"
                            label="Value"
                            name="value"
                            placeholder=""
                            required=""
                            type="text"
                            value=""
                          />
                          
                        </span>
                      </div>
                      <div
                        class="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr form-group"
                      >
                        <span>
                          <div
                            class="checkbox"
                          >
                            <label
                              title=""
                            >
                              <input
                                label="Inverted"
                                name="inverted"
                                placeholder=""
                                type="checkbox"
                                value=""
                              />
                              Inverted
                            </label>
                          </div>
                          
                        </span>
                      </div>
                      <div
                        class="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr form-group"
                      >
                        <label
                          class="ControlLabel-sc-1edmum5-0 hCskeX control-label"
                          for="Description"
                        >
                          Description (optional)
                        </label>
                        <span>
                          <textarea
                            class="FormControl-sc-1amoaox-0 fdBWEX form-control"
                            id="Description"
                            label="Description (optional)"
                            name="description"
                            placeholder=""
                            type="textarea"
                          />
                          
                        </span>
                      </div>
                      <p>
                        <strong>
                          Result:
                        </strong>
                         
                        <span>
                          <em>
                            &lt;empty&gt;
                          </em>
                           
                          <strong>
                            must 
                          </strong>
                          match exactly
                           
                          <em>
                            &lt;empty&gt;
                          </em>
                        </span>
                      </p>
                    </div>
                    <div
                      class="col-md-4"
                    >
                      <div
                        class="Well-sc-19gs1oc-0 eWEWxV matcher-github well well-sm"
                      >
                        The server will try to convert to strings or numbers based on the matcher type as well as it can.
                        <br />
                        <br />
                        <div />
                        <a
                          href="https://github.com/Graylog2/graylog2-server/tree/MAJOR_AND_MINOR_VERSION_MOCK/graylog2-server/src/main/java/org/graylog2/streams/matchers"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                           Take a look at the matcher code on GitHub
                        </a>
                        <br />
                        <br />
                        Regular expressions use Java syntax. 
                        <a
                          href="https://docs.graylog.org/en/MAJOR_AND_MINOR_VERSION_MOCK/pages/streams.html"
                          target="_blank"
                          title="More information"
                        >
                          <div />
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="modal-footer"
              >
                <button
                  class="Button__StyledButton-c9cbmb-0 jVYrA btn btn-default"
                  type="button"
                >
                  Cancel
                </button>
                <button
                  class="Button__StyledButton-c9cbmb-0 jVYrA btn btn-primary"
                  type="submit"
                >
                  Save
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`StreamRuleForm should render an simple StreamRuleForm 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    style="padding-right: 0px;"
  >
    <div
      aria-hidden="true"
    />
    <div
      aria-hidden="true"
      role="dialog"
    >
      <div
        class="fade modal-backdrop in"
      />
      <div
        class="fade Modal-nuree8-0 UjHmV in modal"
        role="dialog"
        style="display: block; padding-right: 0px;"
        tabindex="-1"
      >
        <div
          class="modal-dialog"
        >
          <div
            class="modal-content"
            role="document"
          >
            <div
              class="modal-header"
            >
              <button
                class="close"
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  Ã—
                </span>
                <span
                  class="sr-only"
                >
                  Close
                </span>
              </button>
              <h4
                class="modal-title"
              >
                Bach
              </h4>
            </div>
            <form
              data-testid="modal-form"
              id="StreamRuleForm"
            >
              <div
                class="modal-body"
              >
                <div
                  class="container-fluid"
                >
                  <div>
                    <div
                      class="col-md-8"
                    >
                      <div>
                        field_1
                      </div>
                      <div
                        class="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr form-group"
                      >
                        <label
                          class="ControlLabel-sc-1edmum5-0 hCskeX control-label"
                          for="Type"
                        >
                          Type
                        </label>
                        <span>
                          <select
                            class="FormControl-sc-1amoaox-0 fdBWEX form-control"
                            id="Type"
                            label="Type"
                            name="type"
                            placeholder=""
                            required=""
                            type="select"
                          >
                            <option
                              value="1"
                            >
                              match exactly
                            </option>
                            <option
                              value="2"
                            >
                              match regular expression
                            </option>
                            <option
                              value="3"
                            >
                              greater than
                            </option>
                            <option
                              value="4"
                            >
                              smaller than
                            </option>
                            <option
                              value="5"
                            >
                              field presence
                            </option>
                            <option
                              value="6"
                            >
                              contain
                            </option>
                            <option
                              value="7"
                            >
                              always match
                            </option>
                            <option
                              value="8"
                            >
                              match input
                            </option>
                          </select>
                          
                        </span>
                      </div>
                      <div
                        class="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr form-group"
                      >
                        <label
                          class="ControlLabel-sc-1edmum5-0 hCskeX control-label"
                          for="Value"
                        >
                          Value
                        </label>
                        <span>
                          <input
                            class="FormControl-sc-1amoaox-0 fdBWEX form-control"
                            id="Value"
                            label="Value"
                            name="value"
                            placeholder=""
                            required=""
                            type="text"
                            value="value_1"
                          />
                          
                        </span>
                      </div>
                      <div
                        class="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr form-group"
                      >
                        <span>
                          <div
                            class="checkbox"
                          >
                            <label
                              title=""
                            >
                              <input
                                label="Inverted"
                                name="inverted"
                                placeholder=""
                                type="checkbox"
                                value=""
                              />
                              Inverted
                            </label>
                          </div>
                          
                        </span>
                      </div>
                      <div
                        class="FormGroup__StyledFormGroup-sc-1wv4cm9-0 ekazNr form-group"
                      >
                        <label
                          class="ControlLabel-sc-1edmum5-0 hCskeX control-label"
                          for="Description"
                        >
                          Description (optional)
                        </label>
                        <span>
                          <textarea
                            class="FormControl-sc-1amoaox-0 fdBWEX form-control"
                            id="Description"
                            label="Description (optional)"
                            name="description"
                            placeholder=""
                            type="textarea"
                          >
                            description
                          </textarea>
                          
                        </span>
                      </div>
                      <p>
                        <strong>
                          Result:
                        </strong>
                         
                        <span>
                          <em>
                            field_1
                          </em>
                           
                          <strong>
                            must 
                          </strong>
                          match exactly
                           
                          <em>
                            value_1
                          </em>
                        </span>
                      </p>
                    </div>
                    <div
                      class="col-md-4"
                    >
                      <div
                        class="Well-sc-19gs1oc-0 eWEWxV matcher-github well well-sm"
                      >
                        The server will try to convert to strings or numbers based on the matcher type as well as it can.
                        <br />
                        <br />
                        <div />
                        <a
                          href="https://github.com/Graylog2/graylog2-server/tree/MAJOR_AND_MINOR_VERSION_MOCK/graylog2-server/src/main/java/org/graylog2/streams/matchers"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                           Take a look at the matcher code on GitHub
                        </a>
                        <br />
                        <br />
                        Regular expressions use Java syntax. 
                        <a
                          href="https://docs.graylog.org/en/MAJOR_AND_MINOR_VERSION_MOCK/pages/streams.html"
                          target="_blank"
                          title="More information"
                        >
                          <div />
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="modal-footer"
              >
                <button
                  class="Button__StyledButton-c9cbmb-0 jVYrA btn btn-default"
                  type="button"
                >
                  Cancel
                </button>
                <button
                  class="Button__StyledButton-c9cbmb-0 jVYrA btn btn-primary"
                  type="submit"
                >
                  Save
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
